### PR TITLE
Proposal for TRAPI 1.2 asyncquery

### PR DESCRIPTION
**Pull request by edeutsch**
New proposal based on discussion 2021-06-03

### CODE REVIEW

**Review by marcdubybroad — APPROVED** [review]

File: TranslatorReasonerAPI.yaml [hunk 1]
```diff
@@ -167,6 +217,36 @@ components:
...
+        it requires a callback property.
+      x-body-name: request_body
+      type: object
+      properties:
+        callback:
+          type: string
+          format: uri
+          pattern: ^https?://
+          description: >-
+            Upon completion, the Response will be POSTed to this URL.
```
Comment (patrickkwang) [inline TranslatorReasonerAPI.yaml:232]:
```suggestion
            Upon completion, this server will send a POST request to the callback
            URL with `Content-Type: application/json` header and request body
            containing a JSON-encoded `Response` object.
```

**Review by patrickkwang — CHANGES_REQUESTED** [review]

**Comment by newgene** [comment]
@edeutsch do we want to support "polling" style asyncquery? The advantage is the client-side can actively check for the status.

**Comment by patrickkwang** [comment]
Can the server POST to the callback URL multiple times? Say if the server generates some results quickly and further results more slowly?

**Comment by edeutsch** [comment]
> @edeutsch do we want to support "polling" style asyncquery? The advantage is the client-side can actively check for the status.

I originally thought polling style would be nicer because it gives the client more control, but in our discussion June 3, we arrived at the conclusion that callback was much simpler, stateless, and what ARS already wants and has implemented. And so for the moment we have decided to go callback only. If there is a groundswell to reverse that decision, let it be known!

**Comment by edeutsch** [comment]
> 
> 
> Can the server POST to the callback URL multiple times? Say if the server generates some results quickly and further results more slowly?

Although this was discussed, I do not think we came to a stated conclusion on this. It is my recollection that @MarkDWilliams said that ARS already allows unlimited calls to the callback URL, true? We should definitely clarify in the docs whether this is explicitly desired and permitted. If yes, could perhaps specify a different Response.status depending on whether the results are interim results or final results.

**Review by patrickkwang — APPROVED** [review]

**Comment by newgene** [comment]
@edeutsch if it's just for ARS purpose, I guess that's fine. But if my TRAPI client is not a web application, just a Python script, I don't really have the existence of a callback API at the client side (unless specifically implement one). This is the case if an ARA or a KP needs to consume another TRAPI endpoint.

An async query is not goanna be stateless anyway, it's matter of whether you want to implement "sessions" at the TRAPI server-side or its client-side. If we make TRAPI server state-less, we are making every client to implement "sessions". I would rather to have it at the server-side.

Having said that, if we decided to go with callback for now to just make ARS work, and worry about "polling" support later, I am fine with it.

**Comment by edeutsch** [comment]
> @edeutsch if it's just for ARS purpose, I guess that's fine. But if my TRAPI client is not a web application, just a Python script, I don't really have the existence of a callback API at the client side (unless specifically implement one). This is the case if an ARA or a KP needs to consume another TRAPI endpoint.

Agreed. I presented this argument at the June 3 call, but it did not carry the day. For ARS and ARAs I think callback will be fine and easier for everyone. For an arbitrary Python script behind a NAT firewall, this is hard. (if there's no NAT, then it seems easy, just set up a listener)

> An async query is not goanna be stateless anyway, it's matter of whether you want to implement "sessions" at the TRAPI server-side or its client-side. If we make TRAPI server state-less, we are making every client to implement "sessions". I would rather to have it at the server-side.

I am not well versed in these things, but it seems to me that a request with a callback qualifies as stateless, while a polling scenario does not. Maybe I'm confused about what stateless means.

> Having said that, if we decided to go with callback for now to just make ARS work, and worry about "polling" support later, I am fine with it.

I think this is the path we are on, exactly.

**Comment by edeutsch** [comment]
@MarkDWilliams do you have a comment on the comment above about *multiple* POSTs to the callback URL? interim results? final results? Already supported? Planned? not?

**Review by vdancik — APPROVED** [review]

**Review by CaseyTa — APPROVED** [review]

**Comment by CaseyTa** [comment]
The x-trapi option already implies that `/asyncquery` isn't required to be implemented, but should we document explicitly that it's not required? Could be in the yaml or in the readme or policies markdowns.

**Review by bill-baumgartner — APPROVED** [review]

**Review by rtroper — APPROVED** [review]

**Review by yakaboskic — APPROVED** [review]

**Review by kshefchek — APPROVED** [review]

**Pull request merged**